Various debug-info improvements, fixes and cleanup

frdeso opened the pull request:
Implement file descriptor caching to reduce file descriptor usage.
Fix: bug in dwarf parsing
Change memory allocation and string functions to glib functions.
Move debug-info code to its own directory

In the fdcache commit, I used a hashing function to hash dev and inode numbers that I found on Stackoverflow with no license. Is it okay?

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eepp/babeltrace/25)
<!-- Reviewable:end -->

frdeso committed:
flt.lttng-utils.debug-info: use glib memory and string functions

- Change use of glibc's malloc() to glib's g_new0() to follow the rest of
  the project.

- Change use of glibc string functions to their glib equivalent.

- Remove uses of GNU extension of conditional expressions with omitted
  operands as it would only be useful if the evaluation of the condition
  had side effects which it doesn't. Let's prefer code simplicity.

[0]:  https://gcc.gnu.org/onlinedocs/gcc-4.2.4/gcc/Conditionals.html

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
Fix: flt.lttng-utils.debug-info: Error in src line reporting

Issue
=====
A `debug-info` component does not print the `src` field of a tracepoint
properly when tracepoint is within a for loop.

  #include "tp.h"
  int main(int argc, char *argv[])
  {
      tracepoint(my_app, empty);
      for(int i = 0; i < 1; i++) {
          tracepoint(my_app, empty);
      }
      return 0;
  }

The problem derives from the DWARF information generated by the
compiler. The compiler may place the DWARF content of the for loop in
`DW_TAG_lexical_block` which the current `debug-info` design doesn't
expect.

Here is a summarized version of DWARF tree containing the first
tracepoint in the code above:
  DW_TAG_subprogram
    DW_TAG_inlined_subroutine
      /* tracepoint callsite info */

Here is a summarized version of the DWARF tree containing the tracepoint
in the for loop:
  DW_TAG_subprogram
    DW_TAG_lexical_block
      DW_TAG_inlined_subroutine
      	/* tracepoint callsite info */

The current implementation doesn't expect the presence of a
`DW_TAG_lexical_block` entry as child of the `DW_TAG_inlined_subroutine`
entry and won't look any further down that branch to find the source
information for the current tracepoint. This results in the component
not finding the line number for such callsite and thus leaving the `src`
field empty. Also, on some occasions, the `src` field would contain the
tracepoint definition location (e.g. "tp.h:12") rather then the actual
callsite of that tracepoint.

Solution
========
When iterating over the Debugging Information Entries, if the current
entry contains the address to resolve _but_ is not a of the type
`DW_TAG_inlined_subroutine` try to iterate over the children of that
entry.

Known drawbacks
===============
None.

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
Cleanup: Move `debug-info` in its own sub-directory

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
flt.lttng-utils.debug-info: Implement file descriptor cache

A `flt.lttng-utils.debug-info` component has to open files in order to
do the address resolving when processing a trace. Those file descriptors
are currently associated with a bin_info structure which are stored in a
hash table for each vpid. If, in the same trace, a binary file is loaded
by more than one process, that binary file will be opened twice. This
can create a file descriptor exhaustion problem on long running traces
if the same UST apps are run repeatedly.

This commit implements a file descriptor caching feature to reuse file
descriptors pointing to the same files.

After creating the file descriptor cache (fdcache), users must use the
following two functions to get and put handles to file descriptors:
  struct fd_handle *fd_cache_get_handle(struct fd_cache *fdc, char *path);
  void fd_cache_put_handle(struct fd_cache *fdc, struct fd_handle *handle);

The fd_handles are reference counted internally and their corresponding
file descriptors are closed when the refcount reaches zero.

Files are compared based on the hashed values of inode and device number
rather than solely on the path. This is needed to make sure that a file
opened previously has not changed. This is more likely to happen when
processing a lttng-live trace where the following scenario can happen
and must be supported:
  1. Binary v1 is executed,
  2. Trace data received by debug-info component,
  3. Handle 1 is taken on the bin_a v1 file,
  4. Addresses resolved using handle 1,
  5. Binary is modified,
  6. Binary v2 is compiled,
  7. Binary v2 is executed,
  8. Trace data received by debug-info component,
  9. Handle 2 is taken on the bin_a v2 file,
  10. Addresses resolved using handle 2.

Limitation
----------
It's still be possible to exhaust file descriptors in traces with a
large number of UST apps with no binary file overlap. If this becomes a
problem, we may consider devising an utility to close least recently
used files and reopening file descriptors for later used.

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

eepp commented on plugins/lttng-utils/debug-info.c:
```diff
@@ -114,7 +114,7 @@ struct debug_info {
 	/*
 	 * Hash table of VPIDs (pointer to int64_t) to
-	 * (struct ctf_proc_debug_infos*); owned by debug_info.
+	 * (struct proc_debug_info_sources*); owned by debug_info.
```
What this has to do with the commit's subject?

frdeso replied:
nothing. it's a typo fix. I will create a commit specific for it.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+}
+
+BT_HIDDEN
+void fd_cache_destroy(struct fd_cache *fdc)
```
Go to end (and return) directly if `fdc` is `NULL`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+	 * must open a new FD for it. This replacement of file is more likely
+	 * to happen with a lttng-live source component.
+	 */
+	key = GUINT_TO_POINTER(hash_uint64_t(statbuf.st_dev) ^
```
This is the result of the hashing function, not the key. The key must have all the data, otherwise key collisions could happen. You can't call `g_hash_table_lookup()` or `g_hash_table_insert()` with this digest.

Create a structure for your key (containing the device and inode numbers), and then, when you call `g_hash_table_new_full()`:

* The key hashing function is this here (make a dedicated hashing function).
* The key compare function is another dedicated function comparing the device numbers and inode numbers of two keys.
* The key destroy function frees your allocated key structure.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+	 * to happen with a lttng-live source component.
+	 */
+	key = GUINT_TO_POINTER(hash_uint64_t(statbuf.st_dev) ^
+			hash_uint64_t(statbuf.st_ino));
```
I'm wondering about the portability of `st_ino` on Windows. Is there an alternative to uniquely identify a file on Windows? Or is there a concept that the file cannot change when currently opened?

frdeso replied:
Yes, it seems that on Windows, you can't delete a file if another process has an handle open for normal I/O or as a memory-mapped file.

> The DeleteFile function fails if an application attempts to delete a file that has other handles open for normal I/O or as a memory-mapped file (FILE_SHARE_DELETE must have been specified when other handles were opened).
>The DeleteFile function marks a file for deletion on close. Therefore, the file deletion does not occur until the last handle to the file is closed. Subsequent calls to CreateFile to open the file fail with ERROR_ACCESS_DENIED.

https://docs.microsoft.com/fr-ca/windows/desktop/api/fileapi/nf-fileapi-deletefilea

eepp commented on plugins/lttng-utils/debug-info/fdcache.h:
```diff
@@ -0,0 +1,51 @@
+#include <inttypes.h>
+
+struct fd_handle {
+	int fd;
```
Could you make a `static inline` function to access this FD?

Also, could you move this whole FD cache API to `fd-cache/fd-cache.c` and `include/babeltrace/fd-cache-internal.h`, and prefix everything "internally public" with `bt_fd_cache` (so your handle becomes `bt_fd_cache_handle`), creating a convenience library? This API seems independent enough to do this.

frdeso replied:
Should this really be `fd-cache-internal.h` rather than `fd-cache.h` if it's used by plugins?

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+		fd_internal->ref_count--;
+	} else {
+		gboolean ret;
+		close(fd_internal->fd.fd);
```
`close()` can fail. Not much we can do about it here, but at least log it with `BT_LOGW_ERRNO()`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+	if (fd_internal->ref_count > 1) {
+		fd_internal->ref_count--;
+	} else {
+		gboolean ret;
```
Newline after declarations.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+
+		int fd = open(path, O_RDONLY);
+		if (fd < 0) {
+			if (errno == ENFILE) {
```
Just use `BT_LOGE_ERRNO()`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+
+	ret = stat(path, &statbuf);
+	if (ret < 0) {
+		if (errno == ENOENT) {
```
Just use `BT_LOGE_ERRNO()`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	if (internal_fd->fd.fd >= 0) {
+		close(internal_fd->fd.fd);
```
`close()` can fail. Not much we can do about it here, but at least log it with `BT_LOGW_ERRNO()`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+		struct fd_handle_internal *internal_fd)
+{
+	if (!internal_fd) {
+		return;
```
Use `end` label with `return;` and `goto end;` here. Just for consistency.

eepp commented on plugins/lttng-utils/debug-info/debug-info.c:
```diff
@@ -723,7 +732,7 @@ void handle_bin_info_event(struct debug_info *debug_info,
 
 	event_get_common_context_signed_integer_field_value(event,
-		VPID_FIELD_NAME, &vpid);
+			VPID_FIELD_NAME, &vpid);
```
Why?

eepp commented on plugins/lttng-utils/debug-info/fdcache.h:
```diff
@@ -0,0 +1,51 @@
+struct fd_cache;
+
+BT_HIDDEN
+struct fd_cache *fd_cache_create(void);
```
No need for dynamic allocation here I guess? Just have `bt_fd_cache_init()` and `bt_fd_cache_fini()` which do not dynamically allocate and have a simple `struct bt_fd_cache fd_cache` in `struct debug_info_msg_iter`.

eepp commented on plugins/lttng-utils/debug-info/fdcache.c:
```diff
@@ -0,0 +1,202 @@
+	g_free(fdc);
+}
+
+/* Using simple hash algorithm found on stackoverflow:
```
```
/*
 * Write here.
 */

eepp commented on plugins/lttng-utils/debug-info/debug-info.c:
```diff
@@ -738,15 +747,14 @@ void handle_bin_info_event(struct debug_info *debug_info,
-		debug_info->comp->arg_debug_dir,
-		debug_info->comp->arg_target_prefix);
+	bin = bin_info_create(debug_info->fd_cache, path, baddr, memsz, is_pic,
+			debug_info->comp->arg_debug_dir,
```
Why?

eepp reviewed (changes requested):
> In the fdcache commit, I used a hashing function to hash dev and inode numbers that I found on Stackoverflow with no license. Is it okay?

I don't think such a trivial piece of code is copyrightable.

Please put the `Fix:` prefix at the beginning of the subject line.

frdeso committed:
Fix: lib: Mismatch in function declarations and definitions

Signed-off-by: Simon Marchi <simon.marchi@efficios.com>
Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
flt.lttng-utils.debug-info: cleanup: only print debug message on error

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
flt.lttng-utils.debug-info: cleanup: remove usage old struct name

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
tests: remove unused automake variable

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
flt.lttng-utils.debug-info: cleanup: create dedicated msg_iter destroy function

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso committed:
flt.lttng-utils.debug-info: Don't copy trace class UUID to output trace class

A `flt.lttng-utils.debug-info` component creates an output trace class
that may be different that of the input trace class, so the UUID should
not be copied. In fact, we don't need a UUID at this point. the sink
component can decide to add one later.

Signed-off-by: Francis Deslauriers <francis.deslauriers@efficios.com>

frdeso commented:
Ready for second round of review

eepp reviewed (changes requested):
> Reviewed 18 of 27 files at r2, 17 of 17 files at r3.
> *[Reviewable](https://reviewable.io/reviews/eepp/babeltrace/25)* status: all files reviewed, 15 unresolved discussions (waiting on @frdeso)

___
*[fd-cache/fd-cache.c, line 75 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl51ZX7_VdgSS5pjyd:-Lcl51ZX7_VdgSS5pjye:b-4t0cz6) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L75)):*
> ```C
> static inline
> uint64_t hash_uint64_t(uint64_t x) {
>     x = (x ^ (x >> 30)) * UINT64_C(0xbf58476d1ce4e5b9);
> ```

Are those spaces?
___
*[fd-cache/fd-cache.c, line 109 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl6Kf53kAq79oIJXaP:-Lcl6Kf65qVGJY6BT7iG:b-1vijnq) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L109)):*
> ```C
> 	int ret = 0;
> 
> 	fdc->cache = g_hash_table_new_full(file_key_hash, file_key_equal,
> ```

Exactly!
___
*[fd-cache/fd-cache.c, line 140 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl6jB7EoP6QbJPwFKY:-Lcl6jB7EoP6QbJPwFKZ:b-r560u7) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L140)):*
> ```C
> 	ret = stat(path, &statbuf);
> 	if (ret < 0) {
> 		BT_LOGE_ERRNO("Failed to stat file", "path=%s", path);
> ```

Use `: path=%s` because this is directly appended after the errno message. See other `BT_LOG*_ERRNO()` usages.
___
*[fd-cache/fd-cache.c, line 150 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl72hR5gxCnCsa5uTT:-Lcl72hR5gxCnCsa5uTU:blan7aj) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L150)):*
> ```C
> 	 * to happen with a lttng-live source component.
> 	 */
> 	struct file_key fk = {
> ```

Declaration at the beginning. Only set fields here.
___
*[fd-cache/fd-cache.c, line 178 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7FAa0IZGPIAXSGK6:-Lcl7FAa0IZGPIAXSGK7:b-sf911o) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L178)):*
> ```C
> 		}
> 
> 		memcpy(file_key, &fk, sizeof(*file_key));
> ```

You can also do:

```c
*file_key = fk;
```
___
*[fd-cache/fd-cache.c, line 196 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7Yyu4izfd46vG8sM:-Lcl7Yyu4izfd46vG8sN:b-yascnh) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L196)):*
> ```C
> 
> error:
> 	fd_cache_handle_internal_destroy(fd_internal);
> ```

I guess you want to set `fd_internal` to `NULL` after this.
___
*[fd-cache/fd-cache.c, line 207 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7gfJ2tWnsDlUXOcR:-Lcl7gfJ2tWnsDlUXOcS:b7xnnek) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L207)):*
> ```C
> 
> 	if (!handle) {
> 		return;
> ```

Just for consistency, can you `goto end` here and have the `end` label return at the end of the function?
___
*[fd-cache/fd-cache.c, line 222 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7prK0Iwyht0Iiuhx:-Lcl7prK0Iwyht0Iiuhy:b-gh2s9h) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L222)):*
> ```C
> 		close_ret = close(fd_internal->fd.fd);
> 		if (close_ret == -1) {
> 			BT_LOGW_ERRNO("Failed to close file descriptor", ":fd=%d",
> ```

Space after `:`.
___
*[include/babeltrace/fd-cache-internal.h, line 33 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl90aq3EfbxvPFS5N_:-Lcl90aq3EfbxvPFS5Na:bvn43bv) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L33)):*
> ```objc
> #include <inttypes.h>
> 
> struct bt_fd_handle {
> ```

I would name this `bd_fd_cache_handle` since it's part of the `bt_fd_cache` API.
___
*[include/babeltrace/fd-cache-internal.h, line 47 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl8KsgB6UZUGsXUvi_:-Lcl8KsgB6UZUGsXUvia:b-msrm4r) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L47)):*
> ```objc
> }
> 
> BT_HIDDEN
> ```

Isn't this in `babeltrace-internal.h`?
___
*[include/babeltrace/fd-cache-internal.h, line 54 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl8wDT8S4U_jw5fzdQ:-Lcl8wDT8S4U_jw5fzdR:b-tlzii1) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L54)):*
> ```objc
> 
> BT_HIDDEN
> struct bt_fd_handle *bt_fd_cache_get_handle(struct bt_fd_cache *fdc, char *path);
> ```

`const char *`?
___
*[tests/plugins/test_bin_info.c, line 70 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclB5sG9LqpDj3yPoAl:-LclB5sG9LqpDj3yPoAm:bko1cdu) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L70)):*
> ```C
> 
> 	ret = bt_fd_cache_init(&fdc);
> 	ok(ret == 0, "bt_fd_cache_init successful");
> ```

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
___
*[tests/plugins/test_bin_info.c, line 123 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBDHU5ANq-Y5fKA1x:-LclBDHU5ANq-Y5fKA1y:bko1cdu) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L123)):*
> ```C
> 
> 	ret = bt_fd_cache_init(&fdc);
> 	ok(ret == 0, "bt_fd_cache_init successful");
> ```

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
___
*[tests/plugins/test_bin_info.c, line 177 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBHVW0nBgPFZKzD7l:-LclBHVW0nBgPFZKzD7m:bko1cdu) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L177)):*
> ```C
> 
> 	ret = bt_fd_cache_init(&fdc);
> 	ok(ret == 0, "bt_fd_cache_init successful");
> ```

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
___
*[tests/plugins/test_bin_info.c, line 223 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBICt0bO7eKRfRdUE:-LclBICt0bO7eKRfRdUF:bko1cdu) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L223)):*
> ```C
> 
> 	ret = bt_fd_cache_init(&fdc);
> 	ok(ret == 0, "bt_fd_cache_init successful");
> ```

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.


<!-- Sent from Reviewable.io -->

frdeso reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eepp/babeltrace/25)* status: 30 of 35 files reviewed, 14 unresolved discussions (waiting on @eepp)

___
*[fd-cache/fd-cache.c, line 75 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl51ZX7_VdgSS5pjyd:-LclZ1bA4Gv5E3s6Rhyp:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L75)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Are those spaces?
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 140 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl6jB7EoP6QbJPwFKY:-LclUTah-b_ViQRDFPuH:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L140)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Use `: path=%s` because this is directly appended after the errno message. See other `BT_LOG*_ERRNO()` usages.
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 150 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl72hR5gxCnCsa5uTT:-LclUZUc2jWVDVQQbQ-M:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L150)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Declaration at the beginning. Only set fields here.
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 178 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7FAa0IZGPIAXSGK6:-LclUihTA9QEh6l1DVsI:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L178)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

You can also do:

```c
*file_key = fk;
```
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 196 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7Yyu4izfd46vG8sM:-LclUn7t5m_RB-5i6j1i:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L196)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I guess you want to set `fd_internal` to `NULL` after this.
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 207 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7gfJ2tWnsDlUXOcR:-LclUuFS-DUR9KEE5tFh:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L207)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Just for consistency, can you `goto end` here and have the `end` label return at the end of the function?
</blockquote></details>

Done.
___
*[fd-cache/fd-cache.c, line 222 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl7prK0Iwyht0Iiuhx:-LclV2Lu796Tphn_hsdH:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/fd-cache/fd-cache.c#L222)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Space after `:`.
</blockquote></details>

Done.
___
*[include/babeltrace/fd-cache-internal.h, line 33 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl90aq3EfbxvPFS5N_:-LclWFq6DNamtMRnUZJ7:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L33)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I would name this `bd_fd_cache_handle` since it's part of the `bt_fd_cache` API.
</blockquote></details>

Done.
___
*[include/babeltrace/fd-cache-internal.h, line 47 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl8KsgB6UZUGsXUvi_:-LclWYfJ-Dy7PgecjUh6:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L47)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

Isn't this in `babeltrace-internal.h`?
</blockquote></details>

Done.
___
*[include/babeltrace/fd-cache-internal.h, line 54 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-Lcl8wDT8S4U_jw5fzdQ:-LclWYyuBWdUkLPhwQ2K:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/include/babeltrace/fd-cache-internal.h#L54)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

`const char *`?
</blockquote></details>

Done.
___
*[tests/plugins/test_bin_info.c, line 70 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclB5sG9LqpDj3yPoAl:-LclYmYe9HErHNBz_5f2:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L70)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
</blockquote></details>

Done.
___
*[tests/plugins/test_bin_info.c, line 123 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBDHU5ANq-Y5fKA1x:-LclYmoq18Ue2vW7u_3F:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L123)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
</blockquote></details>

Done.
___
*[tests/plugins/test_bin_info.c, line 177 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBHVW0nBgPFZKzD7l:-LclYnDxB6Yy100G43yz:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L177)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
</blockquote></details>

Done.
___
*[tests/plugins/test_bin_info.c, line 223 at r3](https://reviewable.io/reviews/eepp/babeltrace/25#-LclBICt0bO7eKRfRdUE:-LclYngx7LYMVgUdOakE:b-896fix) ([raw file](https://github.com/eepp/babeltrace/blob/3d823be5be56d5537638b3a7efd73ae5abdc6bd1/tests/plugins/test_bin_info.c#L223)):*
<details><summary><i>Previously, eepp (Philippe Proulx) wrote…</i></summary><blockquote>

I would just `BT_ASSERT()` here since this test is not testing the FD cache library.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

eepp reviewed (approved):
> Reviewed 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/eepp/babeltrace/25)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
